source/loader/ur_print.cpp

```diff
@@ -0,0 +1,3706 @@
+/*
+ *
+ * Copyright (C) 2023 Intel Corporation
+ *
+ * Part of the Unified-Runtime Project, under the Apache License v2.0 with LLVM Exceptions.
+ * See LICENSE.TXT
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ *
+ * @file ur_print.cpp
+ *
+ */
+
+#include "ur_print.h"
+#include "ur_print.hpp"
+
+#include <algorithm>
+#include <sstream>
+#include <string.h>
+
+ur_result_t str_copy(std::stringstream *ss, char *buff, const size_t buff_size,
+                     size_t *out_size) {
+    size_t c_str_size = strlen(ss->str().c_str()) + 1;
+    if (out_size) {
+        *out_size = c_str_size;
+    }
+    if (buff_size < c_str_size) {
+        return UR_RESULT_ERROR_INVALID_SIZE;
+    }
+#if defined(_WIN32)
+    strncpy_s(buff, buff_size, ss->str().c_str(), c_str_size);
+#else
+    strncpy(buff, ss->str().c_str(), std::min(buff_size, c_str_size));
+#endif
+    return UR_RESULT_SUCCESS;
+}
+
+ur_result_t urPrintFunction(enum ur_function_t value, char *buffer,
+                            const size_t buff_size, size_t *out_size) {
+    if (!buffer) {
```

Comment:
you need to allow buffer to be null, otherwise there's no way to get `out_size` without preallocating something.